hack/externalnode/install-vm.sh

```diff
@@ -141,17 +153,280 @@ EOF
+            echoerr "Error docker-compose couldn't be downloaded, exiting"
+            exit 2
+        fi
+    fi
+    set -eo pipefail
+}
+
+# Create docker-compose file with OVS and antrea-agent container specific parameters.
+create_docker_compose_file() {
+    cat << EOF > $AGENT_CONF_PATH/docker-compose.yaml
+version: "3"
+
+services:
+  ovs:
+    image: $ANTREA_DOCKER_IMAGE
+    container_name: ovs
+    volumes:
+        - /var/log/openvswitch:/var/log/openvswitch
+        - /var/run/openvswitch:/var/run/openvswitch
+        - /lib/modules:/lib/modules
+    command: $OVS_BIN
+    network_mode: host
+    cap_add:
+        - NET_ADMIN
+        - SYS_NICE
+        - SYS_ADMIN
+        - IPC_LOCK
+
+  antrea-agent:
+    image: $ANTREA_DOCKER_IMAGE
+    container_name: antrea-agent
+    volumes:
+        - /etc/antrea/:/etc/antrea
+        - /var/run/antrea:/var/run/antrea
+        - /var/run/openvswitch:/var/run/openvswitch
+        - /var/log/antrea:/var/log/antrea
+    environment:
+        - NODE_NAME=$NODE_NAME
+    command: $ANTREA_AGENT_BIN --config $AGENT_CONF_PATH/antrea-agent.conf --logtostderr=false --log_file=$AGENT_LOG_DIR/antrea-agent.log
+    network_mode: host
+    privileged: true
+    cap_add:
+        - SYS_MODULE
+EOF
+}
+
+create_antrea_agent_service_file() {
+    cat >$AGENT_CONF_PATH/antrea-agent.sh << 'EOF'
+#!/bin/bash
+
+AGENT_CONF_PATH="/etc/antrea"
+AGENT_LOG_DIR="/var/log/antrea"
+PIDFILE="$AGENT_CONF_PATH/antrea-agent.pid"
+DOCKER_COMPOSE_YAML="$AGENT_CONF_PATH/docker-compose.yaml"
+COUNT=0
+LIMIT=1000
+
+declare -a OVS_KERNEL_MODULES=("openvswitch")
+declare -a CONTAINERS=("ovs" "antrea-agent")
+
+remove_pidfile() {
+    rm -f "$PIDFILE"
+}
+
+load_ovs_modules() {
+    for module in "${OVS_KERNEL_MODULES[@]}"; do
+        lsmod | grep -q $module
+        if [[ $? -ne 0 ]]; then
+            echo "Load $module kernel module"
+            modprobe $module
+            if [ "$?" -ne 0 ]; then
+                echoerr "Error loading kernel module $module"
+            fi
+        fi
+    done
+}
+
+start_containers() {
+    while [ $COUNT -lt $LIMIT ] ; do
+        for service in "${CONTAINERS[@]}"; do
+            # Check if containers are UP
+            if [ ! "$(docker ps -q -f name=$service)" ]; then
+                if [ "$(docker ps -aq -f status=exited -f name=$service)" ] ; then
+                    echo "$service container is in exited state"
+                fi
+                echo "Starting containers ${CONTAINERS[*]}"
+                # Start both antrea-agent and OVS containers if any exited.
+                docker-compose -f $DOCKER_COMPOSE_YAML up -d
+                ((COUNT=COUNT+1))
+                break
+            fi
+        done
+        sleep 5
+    done
+}
+
+# To validate any other instance of this script is running on this machine.
+if [ -f "$PIDFILE" ]; then
+    # Kills the another instance of this script process.
+    kill -0 "$(cat $PIDFILE)" && kill -9 "$(cat $PIDFILE)"
+fi
```

Comment:
This may be one reason why keeping pid in /etc/antrea is bad: After reboot, this may kill a wrong process. Using /var/run which is a tmpfs doesn't have the problem.

Reply:
Done.